File: 2022/day3/javascript/part1.mjs
import { readFileSync } from 'node:fs';

const input = readFileSync('../input.txt', 'utf-8');

const itemLists = input.split(/\r?\n/);

let prioritySum = 0;

const lowerAlphaAscii = 'a'.charCodeAt(0) - 1;
const upperAlphaAscii = 'A'.charCodeAt(0) - 27;

for (const itemList of itemLists) {
  const halfLen = itemList.length / 2,
    firstCompartment = itemList.slice(0, halfLen),
    secondCompartment = itemList.slice(-halfLen);

  for (let i = 0; i < firstCompartment.length; i++) {
    const item = firstCompartment[i];

    if (secondCompartment.includes(item)) {
      const asciiValue = item.charCodeAt(0);

      prioritySum +=
        asciiValue -
        (asciiValue < lowerAlphaAscii ? upperAlphaAscii : lowerAlphaAscii);
      break;
    }
  }
}

console.log(prioritySum);

File: 2022/day3/c++/part1.cpp
#include <fstream>
#include <iostream>
#include <string>

using namespace std;

const int lower_alpha_ascii = 'a' - 1, upper_alpha_ascii = 'A' - 27;

int main() {
  ifstream input("../input.txt");

  if (!input.is_open()) {
    cerr << "Unable to open input file\n";

    return 1;
  }

  size_t priority_sum = 0;

  string item_list;
  while (getline(input, item_list)) {
    const int half_len = item_list.length() / 2;
    const string first_compartment = item_list.substr(0, half_len),
                 second_compartment = item_list.substr(half_len, half_len);

    for (int i = 0; i < first_compartment.length(); i++) {
      const char item = first_compartment[i];

      if (second_compartment.find(item) != string::npos) {
        priority_sum += item - (item < lower_alpha_ascii ? upper_alpha_ascii
                                                         : lower_alpha_ascii);
        break;
      }
    }
  }

  input.close();

  cout << priority_sum << '\n';
}
